dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,203 @@
+x86 already contains
+[dummy implementation](https://github.com/betrusted-io/xous-core/blob/30b82b25b100e958790973c129dc49e1acca79ec/kernel/src/arch/x86_64.rs)
+which probably will be extended someday.
+
+
+#### Running in DLME
+
+Xous currently cannot run on x86, and DLME on RISC-V is not covered here.
+
+#### Fobnail integration
+
+Xous has network support but currently lacks USB host drivers and TPM support.
+Also, it lacks C interface for calling the kernel and lacks C library. Getting
+[fobnail-attester](https://github.com/fobnail/fobnail-attester) running may
+require a significant amount of work. Also, Xous needs to gain kexec-like
+abilities to chainload target OS.
+
+### seL4
+
+seL4 is a secure L4 family microkernel written in C. It has strong security
+guarantees assured by [Formal proofs](https://sel4.systems/Info/FAQ/proof.pml),
+however these proofs are still incomplete for x86, see
+[Supported Platforms](https://docs.sel4.systems/Hardware) for an up-to-date
+verification status. seL4 due to its microkernel nature provides higher
+isolation, breach in one of the components (like USB driver, network stack)
+wouldn't compromise entire OS contrary to monolithic kernels.
+
+#### Running in DLME
+
+seL4 is a secure L4 family microkernel written in C. It has strong security
+guarantees assured by [Formal proofs](https://sel4.systems/Info/FAQ/proof.pml).
+However, these proofs are still incomplete for x86, see
+[Supported Platforms](https://docs.sel4.systems/Hardware) for an up-to-date
+verification status. seL4, due to its microkernel nature, provides higher
+isolation. A breach in one of the components (like USB driver, network stack)
+wouldn't compromise the entire OS, contrary to monolithic kernels.
+
+#### Fobnail integration
+
+seL4 provides virtually no drivers, except a few drivers listed here.
+According to the page linked above, `libusbdrivers` is inactive and lacks XHCI
+support. seL4 has basic support for an old version of musl C library (v1.1.16).
+
+Using seL4 would require a significant amount of work:
+- Extending USB drivers
+- Implementing USB EEM driver
+- Implementing TPM driver
+- Porting [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+- seL4 runs all processes in unprivileged mode, so the kernel itself would have
+  to be modified to allow booting of another kernel
+
+[Genode](https://github.com/genodelabs/genode) can run on seL4 and provides
+drivers, at least for some devices and a network stack. Its support for seL4
+used to be incomplete, and many components were broken. However, that might have
+changed, and Genode may be an easier way to get software running on seL4.
+
+### Linux
+
+In the case of Linux, a minimal distribution will be prepared that meets the
+requirements of the project. [Yocto Project](https://www.yoctoproject.org/) will
+be used for this, because it gives a lot of freedom in manipulating the elements
+that make up the target system.
```

Comment:
I would appreciate comparative analysis based on available user space.

Reply:
@pietrushnic IIUC you would like an analysis between

* Heads (busybox user space)
* LinuxBoot (Go user space)
* Linux + Yocto built user space
* https://github.com/littlekernel/lk
* https://nv-tegra.nvidia.com/r/admin/repos/3rdparty/ote_partner/tlk

Reply:
@TomaszAIR no, Heads, LinuxBoot and Linux+Yocto are essentially the same but something else built its user space. I think it does matter.

lk and tlk are different thing, @dpsmith may give here a reason behind that, but essentially those kernels are shipped massively, so it means good marketing exposure, potential support of big org, big community backed by commercial vendor etc.

Reply:
lk, tlk seem to be Zephyr competitions in our analysis. I think Zephyr is good since it is base for LibreBMC and new EC for Intel platforms.

Reply:
To clarify, lk is what AOSP/Android bootloader is based on and fixes tend to flow back up to lk. Whereas tlk is a fork of lk by Nvidia to be used as the TEE OS on Arm platform. Google created a derivative of lk which everyone knows as Fuchsia, is a capability-based general operating system. IMHO would consider Fuchsia more of a competitor to Zephyr, while lk is a more lower-level kernel kit.

Reply:
@TomaszAIR Can you include seL4 + CAmkES and/or seL4 Core Platform?

Reply:
> I think Zephyr is good since it is base for LibreBMC and new EC for Intel platforms.

Sorry but I fail to see the connection. Do we plan to attest BMC or EC with Fobnail?

Reply:
@DemiMarie we will include seL4 + CAmkES and seL4 Core Platform. Thank you for your input

Reply:
@dpsmith I added LK and Fuchsia sections, would you take a look at that?

Reply:
@DemiMarie I added CAmkES section, would you take a look at that?
I focused on CAmkES abilities to run VMs. CAmkES doesn't provide any drivers by itself, so the main advantage would be easier component writing.

seL4 Core Platform lags behind CAmkES in terms of completenes. sCP currently has no x86 support, also development is slow, if not halted.

Reply:
@DemiMarie Would take a look at that? Can we close this thread?

Reply:
Sure

Reply:
Great, thanks